library/std/src/sys/sync/mutex/pi_futex.rs

```diff
@@ -0,0 +1,81 @@
+use crate::sync::atomic::AtomicU32;
+use crate::sync::atomic::Ordering::{Acquire, Relaxed, Release};
+use crate::sys::pi_futex as pi;
+
+pub struct Mutex {
+    futex: AtomicU32,
+}
+
+impl Mutex {
+    #[inline]
+    pub const fn new() -> Self {
+        Self { futex: AtomicU32::new(pi::unlocked()) }
+    }
+
+    #[inline]
+    pub fn try_lock(&self) -> bool {
+        self.futex.compare_exchange(pi::unlocked(), pi::locked(), Acquire, Relaxed).is_ok()
+    }
+
+    #[inline]
+    pub fn lock(&self) {
+        if self.futex.compare_exchange(pi::unlocked(), pi::locked(), Acquire, Relaxed).is_err() {
+            self.lock_contended();
+        }
+    }
+
+    #[cold]
+    fn lock_contended(&self) {
+        // Spin first to speed things up if the lock is released quickly.
+        let state = self.spin();
+
+        // If it's unlocked now, attempt to take the lock.
+        if state == pi::unlocked() {
+            if self.try_lock() {
+                return;
+            }
+        };
+
+        pi::futex_lock(&self.futex).expect("failed to lock mutex");
+
+        let state = self.futex.load(Relaxed);
+        if pi::is_owned_died(state) {
+            panic!("failed to lock mutex because its owner is died");
```

Comment:
s/is died/is dead/ or s/is died/died/ would be more correct. I think the following would be clearer though:

```suggestion
            panic!("failed to lock mutex because the thread owning it finished without unlocking it");
```